### PR TITLE
performance: enhance-apis-by-prefetch-and-decorator

### PR DESCRIPTION
**Pull request by W-Shih**
- feature: add decorators to avoid duplicate codes and improve the readability
- performance: enhance tweet apis by prefetch
- performance: enhance-apis-by-decorator
- test: React to decorator enhancement

### CODE REVIEW

File: utils/decorators.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+from rest_framework.response import Response
+
+
+def required_params(method='GET', params=None, is_required_all=True):
+    # <Wayne Shih> 27-Nov-2021
+    # Good practice is not to set a mutable arg as default value
+    if params is None:
+        params = []
+
+    def decorator(view_method):
```
Comment (wen-wayne-shih) [inline utils/decorators.py:25]:
```suggestion
    @wraps(func)
    def decorator(view_method):
```

File: utils/decorators.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                errors_str = f'Request is missing param(s): {missing_params_str}. ' \
+                             f'All missing params are required to provide.'
+                return Response({
+                    'success': False,
+                    'message': 'Please check input.',
+                    'errors': errors_str,
+                }, status=status.HTTP_400_BAD_REQUEST)
+
+            # <Wayne Shih> 27-Nov-2021
+            # if one param is required
```
Comment (wen-wayne-shih) [inline utils/decorators.py:51]:
```suggestion
            # if at least one param is required
```

**Review by wen-wayne-shih — CHANGES_REQUESTED** [review]
must be addressed before merging

File: utils/decorators.py (hunk 1 above)
Comment (W-Shih) [reply utils/decorators.py:27]: resolved as per suggestion

File: utils/decorators.py (hunk 2 above)
Comment (W-Shih) [reply utils/decorators.py]: resolved as per suggestion

**Review by wen-wayne-shih — CHANGES_REQUESTED** [review]
squash before merging

**Review by wen-wayne-shih — APPROVED** [review]

**Pull request merged**